lib/eventmachine.rb

```diff
@@ -960,22 +982,22 @@ def self.connection_count
   end
 
   # The is the responder for the loopback-signalled event.
-  # It can be fired either by code running on a separate thread ({EventMachine.defer}) or on
-  # the main thread ({EventMachine.next_tick}).
-  # It will often happen that a next_tick handler will reschedule itself. We
-  # consume a copy of the tick queue so that tick events scheduled by tick events
+  # Loopbacks are signalled by EventMachine.defer when it has results to post back, or from
+  # the main thread when EventMachine.next_tick is called.
+  # It will often happen that a next_tick handler will reschedule itself. We only work
+  # the tick queue entries that were there when we started; new tick events scheduled by tick events
   # have to wait for the next pass through the reactor core.
   #
   # @private
   def self.run_deferred_callbacks
     until (@resultqueue ||= []).empty?
-      result,cback = @resultqueue.pop
-      cback.call result if cback
+      result, cback = @resultqueue.pop
+      cback.call(result)
```

Comment:
I changed this contract to only put the callbacks on the queue when they're non-nil.